added GET, PATCH, DELETE /shelters/{shelterID} and modified inputValidation class

dotte-osu opened the pull request.

dotte-osu committed:
modified inputValidation

dotte-osu commented on src/utils/tools.js:
```diff
@@ -22,36 +22,63 @@ class SqlUtil {
    * @return {boolean} If Json includes null, return true.
    */
-  static includesNullorEmpty(json, exceptions = []) {
+  static checkNullorEmpty(res, json, exceptions = []) {
```
- changed the function name
- return status code and error message

xHoudek replied:
One potential problem I see with bringing in the error handling here, is that these functions will only ever be able to be used on the controllers. Is there any situation where they might be useful in a model?

These changes definitely make it more convenient. I am just worried about modularity. What do you think?

EDIT: I saw your standup post and saw you had similar concerns. I think we can probably leave this now if it works, and refactor later if we need to

dotte-osu replied:
Yeah I agree, it would be nice to have these functions available (other than controllers). 
I think I figured out the proper way and I changed it some. Reference: https://scoutapm.com/blog/express-error-handling

dotte-osu commented on src/utils/tools.js:
```diff
@@ -22,36 +22,63 @@ class SqlUtil {
    */
-  static hasAllAttrs(json, attrs) {
-    var arr = attrs
+  static checkAttrs(res, json, attrs) {
```
- changed the function name
- return status code and error message

dotte-osu committed:
fixed comment

dotte-osu commented:
I went ahead and fixed inputValidation to return status code and error message.

dotte-osu committed:
added GET, PATCH, DELETE /shelters/{shelterID}

dotte-osu commented:
added GET, PATCH, DELETE /shelters/{shelterID} and modified inputValidation class

xHoudek commented on src/controllers/sheltersController.js:
```diff
@@ -1,47 +1,102 @@
   //check null values
-  else if (inputValidation.includesNullorEmpty(req.body, ["website"])) res.status(400).send("Please provide required input values");
-  else {
+  if (success) success = inputValidation.checkNullorEmpty(res, req.body, ["website"]);
```
Are there any exceptions that come up in the API when this is encountered? It seems like `checkAttrs` will res.send an error message, but then this controller will keep moving forward, which will cause an exception down the line. Does this need to be an `else if`? Or am I misunderstanding?

xHoudek commented on src/controllers/sheltersController.js:
```diff
@@ -1,47 +1,102 @@
   }
+};
 
+const getShelter = async (req, res, next) => {
```
Does this need input validation at all?

xHoudek replied:
Nevermind. Just checked Confluence. This looks good!

xHoudek commented on src/controllers/sheltersController.js:
```diff
@@ -1,47 +1,102 @@
+
+  if (success) {
+    await updateShelterByID(req.params.shelterID, req.body)
+      .then(() => {
```
should this have a `dbResponse.affectedRows` check similar to lines 82 and 83?

xHoudek commented on src/models/sheltersModel.js:
```diff
@@ -39,7 +39,67 @@ const createShelters = async (params) => {
+  var original = await getShelterByID(shelterID);
+  if(original.length == 0) throw new Error('Cannot find shelter data where shelterID=' + shelterID);;
+  //use coalesce() to avoid 'Column ... cannot be null' error 
+  const sql = 'UPDATE Shelter SET ShelterName = coalesce(?, ?), Address = coalesce(?, ?), EmailAddress = coalesce(?, ?), '
```
This is very clever and I've never seen this before. Great way to handle PATCH!

xHoudek reviewed (changes requested):
Just a general comment: all instances of `var` should be replaced with either `let` or `const`.

I appreciate the addition of the tools and tests. Thank you for that!

dotte-osu committed:
fixed tools and error handling

dotte-osu committed:
added newline

dotte-osu committed:
removed async

dotte-osu commented:
- Added error.js and fixed tool.js for better error handling
- Replaced all `var` with `let` (sorry! I use C# daily and I forget about `let`!
-  Updated unit tests
